# Time Series Forecasting — Modeling & Evaluation

## Objectives

In this notebook I will train and evaluate two models on the processed data from `02_feature_engineering.ipynb`: I will use a naive baseline (`lag_1`-as-prediction, `lag_52`-as-prediction) as a floor, Linear Regression as an interpretable baseline, and Random Forest to capture non-linear relationships such as holiday proximity effects/interactions between context variables. 

Performance is evaluated using weighted mean absolute error (WMAE) instead of plain MAE/RMSE to deal with holiday-week business stakes, where weight = 5 if the week is a holiday week, 1 otherwise. 

## Output
- Trained Linear Regression and Random Forest models
- Recommended model with performance metrics
- Justification for model choice and buisness relevance
- Visualizations for README

## 3.1 Setup & Imports
Importing libraries (`numpy`, `pandas`, `matplotlib`, `seaborn`, `sklearn`) for plotting, modeling, and evaluation, and setting absolute paths for reproducibility. Imported `sklearn`'s `mean_absolute_error` since it has a `sample_weight` argument which is needed for WMAE evaluation 

In [19]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# modeling
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor

# evaluation 
from sklearn.metrics import mean_absolute_error

RANDOM_STATE = 42
TRAIN = '../data/processed/train.csv'
TEST = '../data/processed/test.csv'

## 3.2 Load Processed Data 
Loading the processed train/test datasets saved in `02_feature_engineering.ipynb`, and re-casting datetime dtypes. Handles the X/y split by dropping both `Weekly_Sales_log`/`Weekly_Sales` for `X_train`/`X_test`, and defining separate `y_train` by model. Additionally I drop `Date` from `X_train/test`, since raw datetime isn't directly usable by the models (its useful signal is captured through engineered calendar and lag features) resulting in an error. `Date` is retained in `train_df`/`test_df` for later residual-over-time analysis.

In [20]:
train_df = pd.read_csv(TRAIN)
test_df = pd.read_csv(TEST)

train_df['Date'] = pd.to_datetime(train_df['Date'])
test_df['Date'] = pd.to_datetime(test_df['Date'])

print(train_df['Date'].dtype, train_df['Date'].dtype)

datetime64[us] datetime64[us]


Confirm dtype is not string/object

In [21]:
X_train = train_df.drop(columns=['Weekly_Sales_log', 'Weekly_Sales', 'Date'])
X_test = test_df.drop(columns=['Weekly_Sales_log', 'Weekly_Sales', 'Date'])
y_test = test_df['Weekly_Sales']

# linear regression
y_train_log = train_df['Weekly_Sales_log']

# random forest 
y_train_raw = train_df['Weekly_Sales']


In [22]:
X_train.shape, X_test.shape, y_test.shape, y_train_log.shape, y_train_raw.shape

((136342, 36), (124741, 36), (124741,), (136342,), (136342,))

`.shape` output confirms the train and test X/y were defined correctly — `X_train/test`/`y_test` having 136342/124741 rows and 36 columns for `train`, `y_train_log/raw` also have the correct row counts. 

## 3.3 Naive Baselines 
I use a naive baseline of `lag_1/52`-as-predictions as a floor to compare the downstream linear and non-linear models. To do this I compare the WMAE of `X_test` of `lag_1/52` with `y_test` (sample weight set using `np.where`, 5 if true, 1 otherwise). 

In [23]:
weight = np.where(X_test['IsHoliday'], 5, 1)
WMAE_lag_1 = mean_absolute_error(y_test, X_test['lag_1'],sample_weight = weight)
WMAE_lag_52 = mean_absolute_error(y_test, X_test['lag_52'],sample_weight = weight)

print(f"WMAE (lag_1 baseline): {WMAE_lag_1}")
print(f"WMAE (lag_52 baseline): {WMAE_lag_52}")

WMAE (lag_1 baseline): 1800.2473974954214
WMAE (lag_52 baseline): 1891.4855952936084


`lag_1` scored better (lower error) than `lag_52` (~1800 vs. ~1891). This makes sense: `lag_1` captures short-term momentum, while `lag_52` is a noisier signal, since a full year allows economic conditions, store-level changes, and pricing to drift between the two comparison points. This suggests week-to-week momentum drives this data more reliably than year-over-year seasonality, and sets `lag_1` (WMAE ~= 1800) as the stronger naive floor — meaning downstream models should be evaluated against this bar, not just against `lag_52`.

## 3.4 Baseline: Linear Regression (on `Weekly_Sales_log`)

I will fit the baseline Linear Regression model on `X_train` and `y_train_log`, since the skew would effect linear modeling, then after predicting, before computing WMAE I will inverse-transform the values to the original raw values using the following operation: `np.style(x) * np.expm1(abs(x))`. Lasty, I will compare the WMAE against the naive floor (`lag_1`, WMAE ~= 1800), and `lag_52` as a secondary reference. 

In [24]:
lr = LinearRegression()
lr.fit(X_train, y_train_log)
y_pred_lr = lr.predict(X_test)
inv = np.sign(y_pred_lr) * np.expm1(abs(y_pred_lr))
LR_WMAE = mean_absolute_error(y_test, inv, sample_weight=weight)

print(f"WMAE (LR): {LR_WMAE}")

WMAE (LR): 73800.66007698548


## 3.5 Random Forest (on `Weekly_Sales`)

## 3.6 Residual Analysis Over Time

## 3.7 Segmentation Analysis 

## 3.8 Model Comparison

## 3.9 Recommended Model + Limitations